content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -0,0 +1,284 @@
+
+## Computed Properties
+
+Properties that needs to be available on the template, need to be declared in the hash returned by the `data` function.
+This ends up being really handy in documenting what shows up in your template.
+Computed properties are declared in the `computed` hash and behave very similarly.
+
+I love computed properties in Ember and was a bit miffed when there wasn't something similar in React.
+With Vue, you declare your computed properties as functions and Vue is smart enough to observe anything you declare within the function.
+So, we go from something like this Ember:
+
+```javascript
+fullName: computed('firstName', 'lastName', function() {
+  return `${this.firstName} ${this.lastName}`;
+})
+```
+
+To this in Vue:
+
+```javascript
+fullName() {
+  return `${this.firstName} ${this.lastName}`;
+}
+```
+
+Computed properties have the same issues in Vue as Ember, in that it won't observe nested keys more than one level deep and it has some issues with observing arrays.
+
+## The Router
+
+![A Router](https://i.imgur.com/xlUWWN7.jpg?1)
+<aside>Photo by <a href="https://www.pexels.com/@rawpixel">Rawpixel on Pexels</a></aside>
+
+The router in Vue feels like a good mix between react-router and the Ember Router.
+Let's just look at a full example of a router in Vue:
+
+```javascript
+import Vue from 'vue';
+import Router from 'vue-router';
+import { auth } from 'utils';
+
+import { NotFound, Dashboard, Login } from 'routes';
+
+Vue.use(Router);
+
+const router = new Router({
+  mode: 'history',
+  routes: [
+    {
+      path: '*',
+      component: NotFound
+    },
+    {
+      path: '/dashboard',
+      name: 'Dashboard',
+      meta: {
+        layout: 'sidebar',
+        requiresAuth: true
+      },
+      component: Dashboard
+    },
+    {
+      path: '/login',
+      name: 'Login',
+      component: Login,
+      // If they were redirected here, don't bother doing our auth check
+      beforeEnter: async (to, from, next) => {
+        if (to.params.redirected) {
+          next();
+        } else {
+          if (auth.currentUser) {
+            next({ name: 'Dashboard' });
+          } else {
+            next();
+          }
+        }
+      }
+    }
+  ]
+});
+
+router.beforeEach(async (to, from, next) => {
+  const { currentUser } = auth;
+
+  // If the page requires authentication
+  // make sure they are logged in
+  // if they aren't, kick them to the login page
+  if (to.meta.requiresAuth) {
+    if (currentUser) {
+      next();
+    } else {
+      next({
+        name: 'Login',
+        params: { redirected: true } // indicate they are redirected
+      });
+    }
+  } else {
+    next(); // business as usual
+  }
+});
+
+export default router;
+```
+
+Much like react-router, we specify the route name and point it to a component that we have imported (here, we store them in a `routes/index.js` file, but they could live anywhere).
```

Comment:
how does this affect async loading / route-splitting?

Reply:
@sbatson5 can vue do route splitting?

Reply:
Yes it can https://vuejs.org/v2/guide/components-dynamic-async.html